fork-management-for-infra-providers/README.md

```diff
@@ -0,0 +1,148 @@
+
+## Current status
+
+### Repositories
+
+We have 2 repositories per operator
+1. Fork of the upstream repository 
+- Examples:
+  - https://github.com/giantswarm/cluster-api
+  - https://github.com/giantswarm/cluster-api-provider-aws
+  - https://github.com/giantswarm/cluster-api-provider-cloud-director
+  - https://github.com/giantswarm/cluster-api-provider-gcp
+
+1. A repository for bundling the operator as an app 
+- Naming: https://github.com/giantswarm/\<fork-name\>-app
+- Examples
+  - https://github.com/giantswarm/cluster-api-app
+  - https://github.com/giantswarm/cluster-api-provider-aws-app
+  - https://github.com/giantswarm/cluster-api-provider-cloud-director-app
+  - https://github.com/giantswarm/cluster-api-provider-gcp-app
+
+### Syncing fork repositories
+
+There is no common way.
+
+In CAPVCD, since we have some diffs between our fork and upstream repo, we have a mechanism as follows:
+- `main` branch of the repo is the same as the upstream one. We sync this branch by clicking `Sync fork` button in the UI.
+- `giantswarm-main` branch of the repo contains our changes. We manually rebase it on top of `main` branch when necessary and open a new PR each time. The developer is supposed to resolve conflicts during rebasing.
+ 
+### Bundling + Customization
+
+There are some differences between repositories but the general structure in app repositories is as follows:
+
+- We have Makefile + Bash scripts to
+  - Check out a specific commit/branch of fork/upstream repo. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L14)
+  - Generate base manifests by running scripts of the core repo
+  - Customize base manifests with kustomization. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L19)
+  - Update manifests under `helm` folder for the app. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L28)
+
+When it is necessary to update the app content, developers are supposed to
+- run make targets manually to update manifests and values in the app repo
+- open a PR in the app repo
+
+### Image Management
+
+For some providers, we retag upstream images with retagger.
+For some providers like CAPVCD, we build/push images manually.
+
+## Design proposals
+
+### 1. Repo management
+
+#### 1.a Separate repositories for fork and app
+
+We can have two repositories per operator. One is only fork of upstream and the other one is for only bundling it as an app.
+
+#### 1.b One repository for fork and app
+
+We can maintain bundling and customization logic in the same repo.
+
+#### Decision
+
+- `1.a` is the current situation.
+- `1.b` will lead to many conflicts with the upstream repository and it will be hard to automate conflict resolution.
```

Comment:
To add few comments here.

1. Fork management where we "simply" want to maintain the controller code fork (CAPI, CAPA, CAPZ, etc) with our code changes to the upstream go code is already challenging enough.

2. Additionally, maintaining app repos which are packaging aforementioned controllers into helm charts is also challenging enough.

These two are two separate concerns and we should not mix them.

Having consistent, scalable and robust fork management (1) across different forks (for controller/operator code) for different providers will be challenging, as those are upstream projects that are maintained by different groups of people, so we will encounter differences across providers.

Having consistent, scalable and robust controller/operator packaging into apps (2) will be maybe even more challenging, as for now we have done this across different teams and in different ways, and even if same team was doing it, it wouldn't be simple, as there are differences in upstream projects. This, controller/operator packaging into apps, should be out of scope of this RFC.